server/src/main/java/org/elasticsearch/search/aggregations/metrics/min/MinAggregator.java

```diff
@@ -117,4 +157,77 @@ public InternalAggregation buildEmptyAggregation() {
     public void doClose() {
         Releasables.close(mins);
     }
+
+
+    /**
+     * Returns a converter for point values if early termination is applicable to
+     * the context or <code>null</code> otherwise.
+     *
+     * @param context The {@link SearchContext} of the aggregation.
+     * @param parent The parent aggregator.
+     * @param config The config for the values source metric.
+     */
+    public static Function<byte[], Number> getPointReaderOrNull(SearchContext context, Aggregator parent,
+                                                                ValuesSourceConfig<ValuesSource.Numeric> config) {
+        if (context.query() != null &&
+                context.query().getClass() != MatchAllDocsQuery.class) {
+            return null;
+        }
+        if (parent != null) {
+            return null;
+        }
+        if (config.fieldContext() != null && config.script() == null) {
+            MappedFieldType fieldType = config.fieldContext().fieldType();
+            if (fieldType == null || fieldType.indexOptions() == IndexOptions.NONE) {
+                return null;
+            }
+            Function<byte[], Number> converter = null;
+            if (fieldType instanceof NumberFieldMapper.NumberFieldType) {
+                converter = ((NumberFieldMapper.NumberFieldType) fieldType)::parsePoint;
+            } else if (fieldType.getClass() == DateFieldMapper.DateFieldType.class) {
+                converter = (in) -> LongPoint.decodeDimension(in, 0);
+            }
+            return converter;
+        }
+        return null;
+    }
+
+    /**
+     * Returns the minimum value indexed in the <code>fieldName</code> field or <code>null</code>
+     * if the value cannot be inferred from the indexed {@link PointValues}.
+     */
+    public static Number findLeafMinValue(LeafReader reader, String fieldName, Function<byte[], Number> converter) throws IOException {
+        final PointValues pointValues = reader.getPointValues(fieldName);
+        if (pointValues == null) {
+            return null;
+        }
+        final Bits liveDocs = reader.getLiveDocs();
+        if (liveDocs == null) {
+            return converter.apply(pointValues.getMinPackedValue());
+        }
+        final Number[] result = new Number[1];
+        try {
+            pointValues.intersect(new PointValues.IntersectVisitor() {
+                @Override
+                public void visit(int docID) {
+                    throw new UnsupportedOperationException();
+                }
+
+                @Override
+                public void visit(int docID, byte[] packedValue) {
+                    if (liveDocs.get(docID)) {
+                        result[0] = converter.apply(packedValue);
+                        // this is the first leaf with a live doc so the value is the minimum for this segment.
+                        throw new CollectionTerminatedException();
```

Comment:
I'd rather use a similar mechanism as `max` and avoid throwing an exception here.

Reply:
For the `max` we can only check leaves that contain the max value of the segment but for `min` we can start from the first leaf and stop whenever we reach a non-deleted document. If we apply the same logic than `max` to avoid the exception we'd be able to find the minimum value only on leaves that contain the real minimum value regardless of deletions. I am fine with changing the logic here but just wanted to outline the implications.